### PR TITLE
Add AnalyzerManager to manage analyzers for lifetime of an analyzer host.

### PR DESCRIPTION
**Pull request by mavasani**
This is the first step towards unifying the IDE and compiler AnalyzerDrivers towards a common event queue based execution model. This change adds a shared AnalyzerManager to manage analyzers, it ensures the following for the lifetime of an analyzer host:

1. DiagnosticAnalyzer.Initialize(AnalysisContext) is invoked only once per-analyzer.
2. DiagnosticAnalyzer.SupportedDiagnostics is invoked only once per-analyzer.
3. CompilationStartAnalyzerActions registered during Initialize are invoked only once per-analyzer per-compilation.

Consumers of the AnalyzerManager API can query it for all the actions to execute for an analyzer for a given compilation, without worrying about session wide vs per-compilation actions and avoiding duplicate callbacks.

Additionally, the execution of Initialize method and CompilationStartActions has been moved to separate thread and computing the effective set of analyzer actions is now an async method; currently it is done synchronously on the primary thread of the compiler in AnalyzerDriver.Initialize method. This addresses a long outstanding issue that John Hamby had raised about avoiding callbacks into user code on the primary thread.

@heejaechang: I have renamed the AnalyzerManager that you recently added in Features layer to WorkspaceAnalyzerManager to avoid a name clash and move some functionality from it into this AnalyzerManager type.

@srivatsn @heejaechang John Hamby @gafter  can you please take a look?

### CODE REVIEW

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        public static readonly AnalyzerManager Default = new AnalyzerManager();
+
+        // This map stores the tasks to compute HostSessionStartAnalysisScope for session wide analyzer actions, i.e. AnalyzerActions registered by analyzer's Initialize method.
+        // These are run only once per every analyzer.
+        private ImmutableDictionary<DiagnosticAnalyzer, Task<HostSessionStartAnalysisScope>> _sessionScopeMap =
+            ImmutableDictionary<DiagnosticAnalyzer, Task<HostSessionStartAnalysisScope>>.Empty;
+
+        // This map stores the tasks to compute HostCompilationStartAnalysisScope for per-compilation analyzer actions, i.e. AnalyzerActions registered by analyzer's CompilationStartActions.
+        // Compilation start actions will get executed once per-each compilation as user might want to return different set of custom actions for each compilation.
+        private readonly ConditionalWeakTable<Compilation, ConcurrentDictionary<DiagnosticAnalyzer, Task<HostCompilationStartAnalysisScope>>> _compilationScopeMap =
```
Comment (mavasani) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs]: This map is needed currently for the current IDE driver which is not based on event queue. Once it is changed to be based on event queue, we can remove this table and instead ensure that CompilationStartActions are invoked during processing of CompilationStartedEvent, which should happen exactly once per compilation,

File: src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs [hunk 2]
```diff
@@ -459,212 +434,13 @@ internal static bool CanHaveExecutableCodeBlock(ISymbol symbol)
             }
         }
 
-        private static void VerifyArguments(
```
Comment (mavasani) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs]: We don't need to verify arguments as AnalyzerDriverHelper.Execute*Actions are internal.

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+            new ConditionalWeakTable<Compilation, ConcurrentDictionary<DiagnosticAnalyzer, Task<HostCompilationStartAnalysisScope>>>();
+
+        /// <summary>
+        /// Cache descriptors for each diagnostic analyzer. We do this since <see cref="DiagnosticAnalyzer.SupportedDiagnostics"/> is
+        /// a property rather than metadata. We expect it to be cheap and immutable, but we can't force them to be so, we cache them
+        /// and ask only once.
+        /// </summary>
+        private readonly ConditionalWeakTable<DiagnosticAnalyzer, IReadOnlyList<DiagnosticDescriptor>> _descriptorCache =
```
Comment (JohnHamby) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs]: Why is _sessionScope a dictionary and _descriptorCache a conditional weak table?

Comment (mavasani) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs]: Good point, only because they were existing maps from two different sources :) I think ConditionalWeakTable is preferable as we can have analyzers coming from per-project AnalyzerReferences which don't need to stay alive past the compilation?

**Comment by JohnHamby** [comment]
To the limit of my ability to understand them via GitHub (which is new enough to me that I'm not comfortable with it), these changes look good to me.

**Comment by mavasani** [comment]
I am going to check in these changes to unblock John from making progress on his prototype, will address any review feedback subsequently.

File: src/Features/Core/Features.csproj [hunk 4]
```diff
@@ -537,7 +537,6 @@
   <ItemGroup>
     <Compile Include="CodeFixes\Suppression\SuppressionCodeAction.cs" />
     <None Include="packages.config" />
-    <PublicAPI Include="PublicAPI.txt" />
```
Comment (srivatsn) [inline src/Features/Core/Features.csproj]: Why is this being removed?

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs [hunk 5]
```diff
@@ -2924,9 +2924,9 @@ public override IEnumerable<ISymbol> GetSymbolsWithName(Func<string, bool> predi
 
         #endregion
 
-        internal override AnalyzerDriver AnalyzerForLanguage(ImmutableArray<DiagnosticAnalyzer> analyzers, AnalyzerOptions options, Func<Exception, DiagnosticAnalyzer, bool> continueOnAnalyzerException, CancellationToken cancellationToken)
+        internal override AnalyzerDriver AnalyzerForLanguage(ImmutableArray<DiagnosticAnalyzer> analyzers, AnalyzerOptions options, AnalyzerManager analyzerManager, Func<Exception, DiagnosticAnalyzer, bool> continueOnAnalyzerException, CancellationToken cancellationToken)
```
Comment (heejaechang) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: is there a reason why this has to be a API in compilation rather than doing

new AnalyzerDriver(Compilation compilation, Analyzers ....)

Comment (heejaechang) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: parameter seems a bit weird. analyzers, analzyerOptions all looks like related to AnalyzerManager.

Comment (heejaechang) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: continueOnAnalyzerException func, can we not pass around random func to change behavior? either we should use hierarchy so certain type of analyzer driver has same exception handling behavior or use composable/strategy pattern to change certain behavior. with random func, it is so easy to lose control what is going on.

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.AllInOne.cs [hunk 6]
```diff
@@ -86,7 +86,7 @@ public void AnalyzerDriverIsSafeAgainstAnalyzerExceptions()
         {
             var compilation = CreateCompilationWithMscorlib45(TestResource.AllInOneCSharpCode, parseOptions: TestOptions.Regular);
             ThrowingDiagnosticAnalyzer<SyntaxKind>.VerifyAnalyzerEngineIsSafeAgainstExceptions(analyzer =>
-                compilation.GetAnalyzerDiagnostics(new[] { analyzer }, null, CodeAnalysis.DiagnosticExtensions.AlwaysCatchAnalyzerExceptions), AnalyzerDriverHelper.DiagnosticId);
+                compilation.GetAnalyzerDiagnostics(new[] { analyzer }, null, CodeAnalysis.DiagnosticExtensions.AlwaysCatchAnalyzerExceptions));
```
Comment (heejaechang) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.AllInOne.cs]: nice that we at least have factory for exception handlers. now question is either new func (lambda) style or old design pattern, not sure what we prefer.

File: src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs [hunk 7]
```diff
@@ -13,76 +13,63 @@ namespace Microsoft.CodeAnalysis.Diagnostics
 {
     internal class AnalyzerDriverHelper
     {
-        internal const string DiagnosticId = "AD0001";
+        private const string DiagnosticId = "AD0001";
         private const string DiagnosticCategory = "Compiler";
```
Comment (heejaechang) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs]: just curious. what makes something to go into this AnalyzerDriverHelper? some of them seems part of AnalyzerDriver and some other seems not related to each other. so just asking.

File: src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs [hunk 8]
```diff
@@ -13,76 +13,63 @@ namespace Microsoft.CodeAnalysis.Diagnostics
...
         /// <param name="continueOnAnalyzerException">Predicate to decide if exceptions from the action should be handled or not.</param>
         /// <param name="cancellationToken">Cancellation token.</param>
         /// <remarks>
         /// Note that this API doesn't execute any <see cref="CompilationStartAnalyzerAction"/> registered by the Initialize invocation.
-        /// Use <see cref="ExecuteCompilationStartActions(AnalyzerActions, DiagnosticAnalyzer, Compilation, AnalyzerOptions, Action{Diagnostic}, Func{Exception, DiagnosticAnalyzer, bool}, CancellationToken)"/> API
+        /// Use <see cref="ExecuteCompilationStartActions(ImmutableArray{CompilationStartAnalyzerAction}, HostCompilationStartAnalysisScope, Compilation, AnalyzerOptions, Action{Diagnostic}, Func{Exception, DiagnosticAnalyzer, bool}, CancellationToken)"/> API
         /// to get execute these actions to get the per-compilation analyzer actions.
         /// </remarks>
-        public static AnalyzerActions GetSessionAnalyzerActions(
+        public static void ExecuteInitializeMethod(
```
Comment (heejaechang) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs]: passing in arguments looks like immutable fields of analyzer driver. why is it taken out from analyzer driver?

File: src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs [hunk 9]
```diff
@@ -13,76 +13,63 @@ namespace Microsoft.CodeAnalysis.Diagnostics
...
         /// <param name="analyzerOptions">Analyzer options.</param>
         /// <param name="addDiagnostic">Delegate to add diagnostics.</param>
         /// <param name="continueOnAnalyzerException">Predicate to decide if exceptions from the action should be handled or not.</param>
         /// <param name="cancellationToken">Cancellation token.</param>
-        public static AnalyzerActions ExecuteCompilationStartActions(
-            AnalyzerActions actions,
-            DiagnosticAnalyzer analyzer,
+        public static void ExecuteCompilationStartActions(
+            ImmutableArray<CompilationStartAnalyzerAction> actions,
+            HostCompilationStartAnalysisScope compilationScope,
```
Comment (heejaechang) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs]: same question here.

File: src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs [hunk 10]
```diff
@@ -459,212 +434,13 @@ internal static bool CanHaveExecutableCodeBlock(ISymbol symbol)
...
-        {
-            if (getKind == null)
-            {
-                throw new ArgumentNullException(nameof(getKind));
-            }
-
-            VerifyArguments(semanticModel, actions, analyzerOptions, addDiagnostic, continueOnAnalyzerException);
-        }
-
         internal static void ExecuteAndCatchIfThrows(DiagnosticAnalyzer analyzer, Action<Diagnostic> addDiagnostic, Func<Exception, DiagnosticAnalyzer, bool> continueOnAnalyzerException, Action analyze, CancellationToken cancellationToken)
```
Comment (heejaechang) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs]: is it possible to execute analyzer outside of analyzer driver?

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+        // This map stores the tasks to compute HostSessionStartAnalysisScope for session wide analyzer actions, i.e. AnalyzerActions registered by analyzer's Initialize method.
+        // These are run only once per every analyzer.
+        private ImmutableDictionary<DiagnosticAnalyzer, Task<HostSessionStartAnalysisScope>> _sessionScopeMap =
+            ImmutableDictionary<DiagnosticAnalyzer, Task<HostSessionStartAnalysisScope>>.Empty;
+
+        // This map stores the tasks to compute HostCompilationStartAnalysisScope for per-compilation analyzer actions, i.e. AnalyzerActions registered by analyzer's CompilationStartActions.
+        // Compilation start actions will get executed once per-each compilation as user might want to return different set of custom actions for each compilation.
+        private readonly ConditionalWeakTable<Compilation, ConcurrentDictionary<DiagnosticAnalyzer, Task<HostCompilationStartAnalysisScope>>> _compilationScopeMap =
+            new ConditionalWeakTable<Compilation, ConcurrentDictionary<DiagnosticAnalyzer, Task<HostCompilationStartAnalysisScope>>>();
```
Comment (heejaechang) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs]: set concurrency level to 2. otherwise, concurrent dictionary allocates too much stuff to reduce contention based on number of threads the dictionary is used. problem is since we use threadpool tasks, it runs on many many different threads and it confuse the dictionary. and 2 concurrent threads changing the dictionary is most likely enough for us.

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs (hunk 3 above)
Comment (heejaechang) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs]: because we can't control lifetime of DiagnosticAnalyzer. if DiagnosticAnalyzer is ever held strongly, there should be something manage lifetime of DiagnosticAnalyzer. like subscribing to AnalyzerReference changed event or something. otherwise, we will never let DiagnosticAnalyzer once created go away. basically memory leaks.

I didn't want to manage it explicitly, that is why I used ConditionalWeakTable.

one question though, what is lifetime expectation for AnalyzerDriver you are creating? I thought its lifetime is same AnalyzerDriver. but sounds like it is not, then.. not sure how you can strongly hold on to DiagnosticAnalyzer.

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+        /// <summary>
+        /// Cache descriptors for each diagnostic analyzer. We do this since <see cref="DiagnosticAnalyzer.SupportedDiagnostics"/> is
+        /// a property rather than metadata. We expect it to be cheap and immutable, but we can't force them to be so, we cache them
+        /// and ask only once.
+        /// </summary>
+        private readonly ConditionalWeakTable<DiagnosticAnalyzer, IReadOnlyList<DiagnosticDescriptor>> _descriptorCache =
+            new ConditionalWeakTable<DiagnosticAnalyzer, IReadOnlyList<DiagnosticDescriptor>>();
+
+        private Task<HostCompilationStartAnalysisScope> GetCompilationAnalysisScopeAsync(
```
Comment (heejaechang) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs]: I feel this method a bit weird since it belong to analyzer manager. but only state it uses that belong to the analyzer manager is _compilationScopeMap. and it gets everything that belong to analyzer driver as argument.

feels like this class is not where this method is belong to. rather it belong to AnalyzerDriver and how data should flow is reversed.

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+                Task.Run(() =>
+                {
+                    var compilationAnalysisScope = new HostCompilationStartAnalysisScope(sessionScope);
+                    AnalyzerDriverHelper.ExecuteCompilationStartActions(sessionScope.CompilationStartActions, compilationAnalysisScope, compilation,
+                        analyzerOptions, addDiagnostic, continueOnAnalyzerException, cancellationToken);
+                    return compilationAnalysisScope;
+                }, cancellationToken));
+        }
+
+        private Task<HostSessionStartAnalysisScope> GetSessionAnalysisScopeAsync(
```
Comment (heejaechang) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs]: same as the other one. also getTask doesnt require any state that belong to analyzerManager. so why not the method accepts getTask rather than gets all arguments that only used for the getTask?

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+
+            return task;
+        }
+
+        /// <summary>
+        /// Get all the analyzer actions to execute for the given analyzer against a given compilation.
+        /// The returned actions include the actions registered during <see cref="DiagnosticAnalyzer.Initialize(AnalysisContext)"/> method as well as
+        /// the actions registered during <see cref="CompilationStartAnalyzerAction"/> for the given compilation.
+        /// </summary>
+        public async Task<AnalyzerActions> GetAnalyzerActionsAsync(
```
Comment (heejaechang) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs]: same question.

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+                return compilationScope.GetAnalyzerActions(analyzer);
+            }
+
+            return sessionScope.GetAnalyzerActions(analyzer);
+        }
+
+        /// <summary>
+        /// Return <see cref="DiagnosticAnalyzer.SupportedDiagnostics"/> of given <paramref name="analyzer"/>.
+        /// </summary>
+        public ImmutableArray<DiagnosticDescriptor> GetSupportedDiagnosticDescriptors(
```
Comment (heejaechang) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs]: for this to be here, its lifetime must be same as host (in VS case, VS session)

otherwise, cache can't do its expected enforcement. (calling GetSupportedDiagnostics only once per analyzer)

if its expected lifetime is same as host, fields can't hold onto diagnostic analyzer. more specifically project diagnostic analyzer. otherwise we have memory leak. is this checked in?

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 16]
```diff
@@ -81,7 +81,7 @@ public abstract partial class Compilation
             }
         }
 
-        internal abstract AnalyzerDriver AnalyzerForLanguage(ImmutableArray<DiagnosticAnalyzer> analyzers, AnalyzerOptions options, Func<Exception, DiagnosticAnalyzer, bool> continueOnAnalyzerException, CancellationToken cancellationToken);
+        internal abstract AnalyzerDriver AnalyzerForLanguage(ImmutableArray<DiagnosticAnalyzer> analyzers, AnalyzerOptions options, AnalyzerManager analyzerManager, Func<Exception, DiagnosticAnalyzer, bool> continueOnAnalyzerException, CancellationToken cancellationToken);
```
Comment (heejaechang) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs]: I might already put this comment, can we put analyzers and option to analyzer manager?

Comment (heejaechang) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs]: I wonder what is others preference, but I really dont like passing around something like continueOnAnalyzerException.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 17]
```diff
@@ -60,11 +61,11 @@ public AsyncQueue<Diagnostic> DiagnosticQueue
...
         /// Initializes the compilation for the analyzer driver.
-        /// It also computes and initializes <see cref="compilationAnalysisScope"/> and <see cref="_symbolActionsByKind"/>.
+        /// It also computes and initializes <see cref="analyzerActions"/> and <see cref="_symbolActionsByKind"/>.
         /// Finally, it initializes and starts the <see cref="_primaryTask"/> for the driver.
         /// </summary>
         /// <remarks>
-        /// NOTE: This method must only be invoked from <see cref="AnalyzerDriver.Create(Compilation, ImmutableArray{DiagnosticAnalyzer}, AnalyzerOptions, out Compilation, CancellationToken)"/>.
+        /// NOTE: This method must only be invoked from <see cref="AnalyzerDriver.Create(Compilation, ImmutableArray{DiagnosticAnalyzer}, AnalyzerOptions, AnalyzerManager, out Compilation, CancellationToken)"/>.
         /// </remarks>
         private void Initialize(Compilation comp, CancellationToken cancellationToken)
```
Comment (heejaechang) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: is this used in IDE side as well or planned to be used? this run this concurrently. we already had this experiment in IDE side and decide generating garbage is too much. and decide to never run analyzers concurrently.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 18]
```diff
@@ -184,16 +190,18 @@ internal static AnalyzerDriver Create(Compilation compilation, ImmutableArray<Di
...
+        /// <param name="analyzerManager">AnalyzerManager to manage analyzers for analyzer host's lifetime.</param>
         /// <param name="cancellationToken">a cancellation token that can be used to abort analysis</param>
         /// <param name="continueOnAnalyzerException">Delegate which is invoked when an analyzer throws an exception.
         /// If a non-null delegate is provided and it returns true, then the exception is handled and converted into a diagnostic and driver continues with other analyzers.
         /// Otherwise if it returns false, then the exception is not handled by the driver.
         /// If null, then the driver always handles the exception.
         /// </param>
-        protected AnalyzerDriver(ImmutableArray<DiagnosticAnalyzer> analyzers, AnalyzerOptions options, CancellationToken cancellationToken, Func<Exception, DiagnosticAnalyzer, bool> continueOnAnalyzerException = null)
+        protected AnalyzerDriver(ImmutableArray<DiagnosticAnalyzer> analyzers, AnalyzerOptions options, AnalyzerManager analyzerManager, CancellationToken cancellationToken, Func<Exception, DiagnosticAnalyzer, bool> continueOnAnalyzerException = null)
```
Comment (heejaechang) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: this really gives impression that AnalyzerDriver started with simple and as we want more behavior change, rather we refactor we added new flags or optional parameter to change behavior.

**Comment by heejaechang** [comment]
@mavasani I added some comments, not sure since I did it after merge, whether it actually send comments to you or not.

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs (hunk 3 above)
Comment (mavasani) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs]: AnalyzerManager is essentially something that lives for lifetime of analyzer host: command line compile or VS session. I will switch the _sessionScope to be a conditional weak table.

File: src/Features/Core/Features.csproj (hunk 4 above)
Comment (mavasani) [inline src/Features/Core/Features.csproj]: Yes this is not intended. Probably an accidental delete on my part of another merge issue. I'll revert this.

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs (hunk 5 above)
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: I think the term AnalyzerManager is probably a misnomer.
The newly added type is basically something that lives for the lifetime of analyzer host and ensures that individual hosts don't need to track if Initialize method/Compilation start actions/Supported Diagnostics etc. have been executed for an analyzer, and if so what are the cached results. 
This type is not intended to be the primary manager of everything related to analyzers, that is a pretty significant refactoring which doesn't buy us any new functionality and is likely going to change with John's work, so I am not going to attempt that here.
Do you have a better name suggestion for this type?

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.AllInOne.cs (hunk 6 above)
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.AllInOne.cs]: I didn't follow this comment, sorry.

File: src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs (hunk 7 above)
Comment (mavasani) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs]: AnalyzerDriverHelper is a type in shared AnalyzerDriver project, which is compiled into both CodeAnalysis.csproj and Features.csproj.
So basically everything that needs to be shared between IDE and compiler driver, such as core Execute*Actions method that just call into the analyzer and handle exceptions are needed by both drivers and lie in AnalyzerDriverHelper.
AnalyzerDriver is on the other hand something tied closely to "compiler" analyzer driver, is tightly coupled with event queue model and a compilation.
We created the shared AnalyzerDriver project to ensure that we don't end up exposing public APIs, just to match current IDE driver's design. Eventually, the goal is to remove this shared project and make everything related to driving analyzer execution move back into AnalyzerDriver, but that needs IDE driver to be event based.
We should surely do a bunch of refactoring suggestions you have given in this review. But I think we should wait for IDE driver to move to event queue before we do those.

File: src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs (hunk 8 above)
Comment (mavasani) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs]: I didn't follow the comment, what arguments are you referring to?

File: src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs (hunk 10 above)
Comment (mavasani) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs]: Right now we have 2 analyzer drivers, which one are you referring to?
This type, AnalyzerDriverHelper, is a shared helper for both drivers, so none of the methods here can move to AnalyzerDriver in CodeAnalysis project, until we switch IDE driver to be event based.

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs (hunk 12 above)
Comment (mavasani) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs]: Again I think there are 2 points of confusion here.

1. AnalyzerManager is not a "manage everything for analyzers" type. It is just a shared helper between IDE and compiler analyzer hosts.
2. AnalyzerDriver that you are referring to is compiler's analyzer driver and can't be consumed by IDE driver, but all the shared functionality in shared AnalyzerDriver project is needed by both drivers.

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs (hunk 13 above)
Comment (mavasani) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs]: same comment as before, we don't want both IDE and compiler driver to provide the delegate to compute task for computing analyzer actions.

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs (hunk 14 above)
Comment (mavasani) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs]: and same answer :)

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs (hunk 15 above)
Comment (mavasani) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs]: Yes, I will switch _sessionScopeMap to be a conditional weak table.

File: src/Compilers/Core/Portable/Compilation/Compilation.cs (hunk 16 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs]: Nope, we should probably rename this type from AnalyzerManager to something else that doesn't give a notion of it being the manager for everything related to analyzers.
Yes, hopefully we won't need to pass around continueOnAnalyzerException once we move IDE driver to be event queue. Until then we need this because IDE driver needs to invoke the shared Execute*Actions methods which needs this parameter.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 17 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: Currently just used by command line compiler.
What is being run concurrently here, are you pointing to different analyzers running concurrently?
If yes, that is the design of compiler analyzer driver, it *will* run analyzers in parallel, but each analyzer's actions will never be run concurrently.

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs (hunk 5 above)
Comment (heejaechang) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: if things going to change, I am okay, but hopely at the end, we do some cleanup.

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.AllInOne.cs (hunk 6 above)
Comment (heejaechang) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.AllInOne.cs]: previously, we would have never passed around func like primary way to change behavior of same code.

either should hierarchy of a type

IDEAnalyzerDriver { override bool ContinueOnException }, CompilerAnalyzerDriver { override bool Continue ...}

or ExcentpionHandlerFactory.CreateCrashOnBuiltInAnalyzerExceptionHandler(), ExceptionHandlerFactory.CreateC.... which return something like IDiagnosticExceptionHandler.

rather than passing around func.

File: src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs (hunk 7 above)
Comment (heejaechang) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs]: :+1:

File: src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs (hunk 8 above)
Comment (heejaechang) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs]: hostCompilationStartAnalysisScope, continueOnAnalyzerexception and etc. but I think you already answered my question from previous comment.

File: src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs (hunk 10 above)
Comment (heejaechang) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs]: yep. understood.

Comment (heejaechang) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs]: still they could have base BaseAnalyzerDriver type that contains all these. but well, like you said, if all these are going to be changed anyway then.

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs (hunk 12 above)
Comment (heejaechang) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs]: okay.

File: src/Compilers/Core/Portable/Compilation/Compilation.cs (hunk 16 above)
Comment (heejaechang) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs]: sounds good

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 17 above)
Comment (heejaechang) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: if eventually we will share AnalyzerDriver, I think we need something that control concurrency. I mean any concurrency such as running two different analyzer concurrently.

like I said in previous comment, we already tried that, and that caused too much GC to happen since allocation rate spike up. IDE side, we should not run anything concurrently except user initiated actions such as fix all or ctrl+.

as you might see from my previous email, just having higher priority queue with no concurrency made us to miss 2-core typing goal. having concurrency in IDE with higher priority queue might make us to even miss 4-core typing goal.

**Comment by heejaechang** [comment]
:+1:

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 17 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: Agreed. I personally believe even forcing the API users to execute all analyzers instead of selectively running analyzers is very important from an API perspective. Well all this can come later though. Probably @JohnHamby might want to make a note of concurrency requirement in IDE here.

Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: I meant *not* forcing execution of all analyzers together.
